third-party/realdds/src/dds-stream-server.cpp

```diff
@@ -29,24 +36,76 @@ dds_stream_server::~dds_stream_server()
 }
 
 
-std::shared_ptr< dds_topic > const& dds_stream_server::get_topic() const
+dds_video_stream_server::dds_video_stream_server( std::string const & name,
+                                                  dds_stream_profiles const & profiles,
+                                                  int default_profile_index )
+    : dds_stream_server( name, profiles, default_profile_index )
 {
+    for( auto & profile : profiles )
+        if( ! std::dynamic_pointer_cast< dds_video_stream_profile >( profile ) )
+            DDS_THROW( runtime_error, "profile '" + profile->to_string() + "' is not a video profile" );
+}
+
+
+dds_motion_stream_server::dds_motion_stream_server( std::string const & name,
+                                                    dds_stream_profiles const & profiles,
+                                                    int default_profile_index )
+    : dds_stream_server( name, profiles, default_profile_index )
+{
+    for( auto & profile : profiles )
+        if( ! std::dynamic_pointer_cast< dds_motion_stream_profile >( profile ) )
+            DDS_THROW( runtime_error, "profile '" + profile->to_string() + "' is not a motion profile" );
+}
+
+
+void dds_video_stream_server::open( std::string const & topic_name, std::shared_ptr< dds_publisher > const & publisher )
+{
+    if( is_open() )
+        DDS_THROW( runtime_error, "stream '" + name() + "' is already open" );
+
+    auto topic = topics::device::image::create_topic( publisher->get_participant(), topic_name.c_str() );
+
+    _writer = std::make_shared< dds_topic_writer >( topic, publisher );
+    _writer->run( dds_topic_writer::writer_qos( BEST_EFFORT_RELIABILITY_QOS ) );  // no retries
+}
+
+
+void dds_motion_stream_server::open( std::string const & topic_name,
+                                     std::shared_ptr< dds_publisher > const & publisher )
+{
+    if( is_open() )
+        DDS_THROW( runtime_error, "stream '" + name() + "' is already open" );
+
+    auto topic = topics::device::image::create_topic( publisher->get_participant(), topic_name.c_str() );
+
+    _writer = std::make_shared< dds_topic_writer >( topic, publisher );
+    _writer->run( dds_topic_writer::writer_qos( BEST_EFFORT_RELIABILITY_QOS ) );  // no retries
+}
+
+
+std::shared_ptr< dds_topic > const & dds_stream_server::get_topic() const
+{
+    if( ! is_open() )
+        DDS_THROW( runtime_error, "stream '" + name() + "' must be open to get_topic()" );
+
     return _writer->topic();
 }
 
 
 void dds_stream_server::start_streaming( const image_header & header )
 {
+    if( ! is_open() )
```

Comment:
Should check `is_streaming()`?

Reply:
I left the logic as it was before I moved to stream-server.
You're likely right, but not for this PR.